core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1211 @@
+          case _ =>
+            false
+        }
+      }
+    }
+
+    final override def equals(that: Any): Boolean = that match {
+      case node: CollisionNode[_, _] =>
+        (this.collisionHash == node.collisionHash) &&
+        (this.contents.length == node.contents.length) &&
+        this.contents.forall(kv => node.contents.exists(_ == kv))
+      case _ =>
+        false
+    }
+
+    final override def toString(): String = {
+      s"""CollisionNode(hash=${collisionHash}, values=${contents.iterator.mkString("[", ",", "]")})"""
+    }
+  }
+
+  /**
+    * A CHAMP bitmap node. Stores key-value pair and node positions in the `contents` array in the `keyValueMap` and
+    * `nodeMap` integer bitmaps respectively.
+    *
+    * The index of an element is calculated from a 5-bit segment of the hash of the key. The segment to use is
+    * determined according to the depth in the structure, starting with the least significant bits at the root level.
+    *
+    * When there are collisions in the 5-bit segment of the hash at the current depth in the structure, a new subnode
+    * must be created in order to store the colliding elements. In this subnode, the next 5-bit segment is used to
+    * determine the order of elements.
+    *
+    * Key-value pairs are stored at consecutive indices in the array, indexed from the start of the array
+    * and ordered according to the relative indices calculated from the hash of the key.
+    *
+    * Sub-nodes are stored at the end of the array, indexed from the end of the array and ordered according
+    * to the relative indices calculated from the hash of their keys. As a result of this indexing method
+    * they are stored in reverse order.
+    *
+    * @tparam K the type of the keys contained in this node.
+    * @tparam V the type of the values contained in this node.
+    * @param keyValueMap integer bitmap indicating the notional positions of key-value elements in the `contents` array.
+    * @param nodeMap integer bitmap indicating the notional positions of node elements in the `contents` array.
+    * @param contents an array of `A` value elements and `Node[A]` sub-node elements.
+    * @param size the number of value elements in this subtree.
+    */
+  final private[HashMap] class BitMapNode[K, +V](
+    val keyValueMap: Int,
+    val nodeMap: Int,
+    val contents: Array[Any],
+    val size: Int
+  )(implicit hashKey: Hash[K])
+      extends Node[K, V] {
+
+    final def hasKeyValues: Boolean =
+      keyValueMap != 0
+
+    final def hasNodes: Boolean =
+      nodeMap != 0
+
+    final def allElementsCount: Int =
+      keyValueCount + nodeCount
+
+    final def keyValueCount: Int =
+      Integer.bitCount(keyValueMap)
+
+    final def nodeCount: Int =
+      Integer.bitCount(nodeMap)
+
+    final private def hasNodeAt(bitPos: Int): Boolean =
+      (nodeMap & bitPos) != 0
+
+    final private def hasKeyValueAt(bitPos: Int): Boolean =
+      (keyValueMap & bitPos) != 0
```

Comment:
I wonder if it might be worth adding an inline annotation to some of these methods